servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/TransportObserverUtils.java

```diff
@@ -0,0 +1,96 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.transport.netty.internal;
+
+import io.servicetalk.transport.api.ConnectionObserver;
+import io.servicetalk.transport.api.TransportObserver;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.util.AttributeKey;
+
+import javax.annotation.Nullable;
+
+import static io.netty.util.AttributeKey.newInstance;
+
+/**
+ * Utilities for {@link TransportObserver}.
+ */
+public final class TransportObserverUtils {
+
+    private static final AttributeKey<ConnectionObserver> CONNECTION_OBSERVER = newInstance("ConnectionObserver");
+    private static final AttributeKey<Throwable> CONNECTION_ERROR = newInstance("ConnectionError");
+
+    private TransportObserverUtils() {
+        // No instances
+    }
+
+    /**
+     * Assigns a {@link ConnectionObserver} to the passed {@link Channel}.
+     *
+     * @param channel a {@link Channel} to assign a {@link ConnectionObserver} to
+     * @param observer a {@link ConnectionObserver}
+     */
+    public static void assignConnectionObserver(final Channel channel, @Nullable final ConnectionObserver observer) {
+        if (observer == null) {
+            return;
+        }
+        channel.attr(CONNECTION_OBSERVER).set(observer);
+        channel.closeFuture().addListener((ChannelFutureListener) future -> {
+            Throwable t = connectionError(channel);
+            if (t == null) {
+                observer.connectionClosed();
+            } else {
+                observer.connectionClosed(t);
+            }
+            channel.attr(CONNECTION_OBSERVER).set(null);
+        });
+    }
+
+    /**
+     * Returns {@link ConnectionObserver} associated with the passed {@link Channel}.
+     *
+     * @param channel {@link Channel} that may have assigned {@link ConnectionObserver}
+     * @return {@link ConnectionObserver} associated with the passed {@link Channel}
+     */
+    @Nullable
+    public static ConnectionObserver connectionObserver(final Channel channel) {
+        return channel.attr(CONNECTION_OBSERVER).get();
+    }
+
+    /**
+     * Assigns a {@link Throwable} to the passed {@link Channel}.
+     *
+     * @param channel a {@link Channel} to assign a {@link Throwable} to
+     * @param error a {@link Throwable}
+     */
+    public static void assignConnectionError(final Channel channel, final Throwable error) {
+        if (connectionObserver(channel) != null) {
+            channel.attr(CONNECTION_ERROR).setIfAbsent(error);
```

Comment:
IIUC, the current code paths registering an error is a subset of things you would want to cover eventually.
Looks like we need to have some association with the close-handler to determine which error should be the cause of closure instead of this "first-writer-wins" semantics?

Reply:
Not sure about `CloseHandler` because a `Channel` may be closed from different places. After investigating different closure cases it looks acceptable to use "first-writer-wins" semantics because channels are always closed right after observing an error.

Reply:
> Not sure about CloseHandler because a Channel may be closed from different places.

Sure close can be initiated from multiple places but close-handler holds the close state machine (for H1). 
Having the determination here that the first error is important and the rest are silently ignored essentially is ignoring the state represented by the close state machine (close handler for h1 and netty for h2).

This change only contains 2 calls to this method, how about we determine the mechanism to handle connection errors as a follow-up when we have a better idea about how to handle the reason for close holistically.